lib/input_sanitizer/v2/types.rb

```diff
@@ -93,8 +93,22 @@ def call(value, options = {})
           raise InputSanitizer::ValueError.new(value, options[:minimum], options[:maximum]) if options[:minimum] && string.length < options[:minimum]
           raise InputSanitizer::ValueError.new(value, options[:minimum], options[:maximum]) if options[:maximum] && string.length > options[:maximum]
         end
+
+        if options[:strip_4byte_chars] && !options[:already_stripped]
+          value_without_4byte_chars = strip_4byte_chars(value)
+          updated_options = options.merge(:already_stripped => true) # to prevent infinite loop
+          call(value_without_4byte_chars, updated_options) # run checks once again to ensure string is still valid after stripping 4-byte chars
+        else
+          value
+        end
       end
     end
+
+    private
+
+    def strip_4byte_chars(string)
+      string.chars.reject { |char| char.bytesize >= 4 }.join
```

Comment:
```
irb(main):011:0> RUBY_VERSION
=> "2.6.3"
irb(main):012:0> 'abcd'.chars.object_id
=> 70139492059380
irb(main):013:0> 'abcd'.chars.object_id
=> 70139492130600
irb(main):014:0> 'abcd'.chars.object_id
=> 70140041841820
```

since this is a shared code / library and we don't know the length of the string that will be sanitized - how about we optimize it a bit

```
irb> "test\u{1F435}".each_char.each_with_object(String.new) { |c, s| s << c if c.bytesize < 4 }
=> "test"
```

each_char is an enumerator so should be better (definitely not worse) in terms of allocations because it goes character by character and then chain it with another enumerator (for readability) and do in-place additions to mutable string. I think that if we format it well it's not so much less readable than your proposition, WDYT?

Reply:
Hm, I'm not sure if I understand your concern. 

`chars` is a shorthand for `each_char.to_a`, so that means that different object IDs means different arrays, right? What are you exactly concerned about? I think it will be allocated once, then it will iterate over this array and reject 4-byte chars (I think without allocating new array on each rejection) 🤔

Reply:
each_chars -> no array alocation = 0 allocatinos + building string
`chars` -> 1 array allocation + new array allocation for `reject` = 2 allocations + building string

Reply:
Allocations cost memory and time - if we can ommit them without major readability loss I'd go for it in code shared in many services